src/Microsoft.AspNet.OData.Shared/Extensions/ODataPathExtensions.cs

```diff
@@ -29,28 +37,69 @@ internal static class ODataPathExtensions
 
                 if (pathWithoutLastSegmentCastType.LastSegment is KeySegment)
                 {
-                    keys = GetKeysFromKeySegment(pathWithoutLastSegmentCastType.LastSegment as KeySegment);
+                    keys = ODataPathHelper.GetKeysFromKeySegment(pathWithoutLastSegmentCastType.LastSegment as KeySegment);
                 }
             }
             // Books(1)/Authors/Namespace.SpecialAuthor/(1000)
             else if (path.LastSegment is KeySegment)
             {
-                keys = GetKeysFromKeySegment(path.LastSegment as KeySegment);
+                keys = ODataPathHelper.GetKeysFromKeySegment(path.LastSegment as KeySegment);
             }
 
             return keys;
         }
 
-        private static Dictionary<string, object> GetKeysFromKeySegment(KeySegment keySegment)
+        /// <summary>
+        /// Return the last segment in the path, which is not a <see cref="TypeSegment"/> or <see cref="KeySegment"/>.
+        /// </summary>
+        /// <param name="path">The <see cref="ODataPath"/>.</param>
+        /// <returns>An <see cref="ODataPathSegment"/>.</returns>
+        public static ODataPathSegment GetLastNonTypeNonKeySegment(this ODataPath path)
         {
-            Dictionary<string, object> keys = new Dictionary<string, object>();
+            if (path == null)
+            {
+                throw Error.ArgumentNull(nameof(path));
+            }
 
-            foreach (KeyValuePair<string, object> kvp in keySegment.Keys)
+            if (path.Count == 1)
             {
-                keys.Add(kvp.Key, kvp.Value);
+                return path.LastSegment;
             }
 
-            return keys;
+            // If the path is Employees(2)/NewFriends(2)/Namespace.MyNewFriend where Namespace.MyNewFriend is a type segment,
+            // This method will return NewFriends NavigationPropertySegment.
+
+            while (path.LastSegment is TypeSegment || path.LastSegment is KeySegment)
+            {
+                if (path.LastSegment is TypeSegment)
+                {
+                    // We remove the last type segment from the path.
+                    // E.g If the path is Employees(2)/NewFriends(2)/Namespace.MyNewFriend where Namespace.MyNewFriend,
```

Comment:
complete this statement..where Namespace.MyNewFried is?

Reply:
Resolved